src/main/java/io/jenkins/plugins/ml/IPythonBuilder.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * The MIT License
+ *
+ * Copyright 2020 Loghi Perinpanayagam.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package io.jenkins.plugins.ml;
+
+import hudson.Extension;
+import hudson.Launcher;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.BuildListener;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.Builder;
+import hudson.util.FormValidation;
+import org.apache.zeppelin.interpreter.InterpreterException;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class IPythonBuilder extends Builder {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(IPythonBuilder.class);
+    private static InterpreterManager interpreterManager;
+    private static IPythonUserConfig userConfig;
+
+    private final String code;
+    private final String filePath;
+    private String parserType;
+
+    @DataBoundConstructor
+    public IPythonBuilder(String code,String filePath, String parserType) {
+        this.code = code;
+        this.filePath = filePath;
+        this.parserType = parserType;
+    }
+
+    public String getCode() {
+        return code;
+    }
+
+    private synchronized InterpreterManager getIPythonInterpreterManager(IPythonUserConfig jobUserConfig) {
+
```

Comment:
While this works, I think we need to think a bit more how to instantiate the interpreter @Loghijiaha .

From what I understood after a cursory look at the PR (didn't have time to open it in the IDE, sorry), the logic goes as follows:

- User configures the interpreter in the global configuration of Jenkins
- There is still no interpreter created, i.e. user is no aware whether it will work or not
- User configures a job in Jenkins with the machine-learning-plugin builder
- There is still no interpreter created
- 10 users submit the job simultaneously
    * at this point, this method might be called 10 times (worst case I think, rare, but there's a chance, right :)
    * the method if synchronized, so the JVM allows only one to acquire the lock and run it
    * at this point, we will try to instantiate the `IPythonInterpreterManager`
    * If it succeeds, all 10 users will be able to use the interpreter
    * If it fails... then the next user in the line to acquire the lock of the method will come in, and then try to instantiate it... and so it goes until all 10 users are done

IMO, we should fail earlier. I will have to do a bit of thinking to be more helpful... but I remember there used to be some sort of lightweight task that could be run in parallel in Jenkins... I had a look at the extension points and the closest I found was the `PeriodicWork`, but that's not what I was looking for.

Had a look then at the Selenium Plugin. A friend used to maintain this plugin, and I remember he started a process to connect to the selenium server and send commands. Have a look later and see if that gives you any ideas: https://github.com/jenkinsci/selenium-plugin/blob/master/src/main/java/hudson/plugins/selenium/PluginImpl.java

Reply:
More to discuss and decide @kinow. I will research this more based on other plugins that have similar functionalities like parrel connection. I will fix some of obvious issues in this PR then we will decide either merge this and then improve continuously or more work on this. I guess for phase 1, we have only left with the file parser. We have enough time to discuss and improve those drawbacks :).

Reply:
Good rationale @Loghijiaha . Better make progress with small steps than block work for too long. Take your time and assess what you think would work best here ;)